frontend/controller/actions/identity.js

```diff
@@ -770,6 +789,115 @@ export default (sbp('sbp/selectors/register', {
+        }
+      })
+    })
+  },
+  'gi.actions/identity/deleteChatRoomReadUntil': ({ contractID, deletedHeight }: {
+    contractID: string, deletedHeight: number
+  }) => {
+    return sbp('okTurtles.eventQueue/queueEvent', CHATROOM_LOGS, () => {
+      return requireToSuccess(async () => {
+        const currentChatRoomLogs = await sbp('gi.actions/identity/loadChatRoomLogs')
+
+        if (currentChatRoomLogs[contractID].readUntil.deletedHeight === undefined) {
+          currentChatRoomLogs[contractID].readUntil['deletedHeight'] = deletedHeight
+          await sbp('gi.actions/identity/saveChatRoomLogs', contractID, currentChatRoomLogs)
+        }
+      })
+    })
+  },
+  'gi.actions/identity/addChatRoomUnreadMessage': ({ contractID, messageHash, type, createdHeight }: {
+    contractID: string, messageHash: string, type: string, createdHeight: number
+  }) => {
+    return sbp('okTurtles.eventQueue/queueEvent', CHATROOM_LOGS, () => {
+      return requireToSuccess(async () => {
+        const currentChatRoomLogs = await sbp('gi.actions/identity/loadChatRoomLogs')
+
+        // NOTE: should ignore to add unreadMessages before joining chatroom
+        if (currentChatRoomLogs[contractID].readUntil) {
+          const index = currentChatRoomLogs[contractID].unreadMessages.findIndex(msg => msg.messageHash === messageHash)
+          if (index < 0) {
+            currentChatRoomLogs[contractID].unreadMessages.push({ messageHash, type, createdHeight })
+            await sbp('gi.actions/identity/saveChatRoomLogs', contractID, currentChatRoomLogs)
+          }
+        }
+      })
+    })
+  },
+  'gi.actions/identity/deleteChatRoomUnreadMessage': ({ contractID, messageHash }: {
+    contractID: string, messageHash: string
+  }) => {
+    return sbp('okTurtles.eventQueue/queueEvent', CHATROOM_LOGS, () => {
+      return requireToSuccess(async () => {
+        const currentChatRoomLogs = await sbp('gi.actions/identity/loadChatRoomLogs')
+
+        // NOTE: should ignore to delete unreadMessages before joining chatroom
+        if (currentChatRoomLogs[contractID].readUntil) {
+          const index = currentChatRoomLogs[contractID].unreadMessages.findIndex(msg => msg.messageHash !== messageHash)
+          if (index >= 0) {
+            currentChatRoomLogs[contractID].unreadMessages.splice(index, 1)
+            await sbp('gi.actions/identity/saveChatRoomLogs', contractID, currentChatRoomLogs)
+          }
+        }
+      })
+    })
+  },
+  'gi.actions/identity/deleteChatRoomLog': ({ contractID }: { contractID: string }) => {
+    return sbp('okTurtles.eventQueue/queueEvent', CHATROOM_LOGS, () => {
+      return requireToSuccess(async () => {
+        const currentChatRoomLogs = await sbp('gi.actions/identity/loadChatRoomLogs')
+        delete currentChatRoomLogs[contractID]
+        await sbp('gi.actions/identity/saveChatRoomLogs', contractID, currentChatRoomLogs)
+      })
+    })
+  },
```

Comment:
Could you please explain what these functions are for? This seems like it's storing a lot of data and I thought the issue was about just how the unread message location is stored.

Would it be possible to store the unread message information just by storing the hash of the oldest unread message?

Reply:
It looks like there are many utility functions to manage chatroom logs. But I don't think it's not that too much, and they are necessary. I will explain the functions one by one.

##### `loadChatRoomLogs` and `saveChatRoomLogs`
These two functions are to load/save chatRoomLogs from/to the backend.

##### `resetChatRoomLogs`
This function is to load chatRoomLogs from the backend and set vuex state with it.

##### `addChatRoomLog` and `deleteChatRoomLog`
These functions are to create(or initialise)/delete logs for one chatroom inside the chatRoomLogs. As I have mentioned before, the chatRoomLogs is in the following format.
```
chatRoomLogs: {
  [chatRoomID]: {
    readUntil: { messageHash, createdHeight, deletedHeight? },
    unreadMessages: [{ messageHash, type, createdHeight }]
  }
}
```
So in initialising, just create a new log for the new chatroom, and set it `{ readUntil: null, unreadMessages: [] }`

##### `setChatRoomReadUntil` and `deleteChatRoomReadUntil`
These two functions are to set/delete the readUntil state inside the chatRoomLog.

##### `addChatRoomUnreadMessage` and `deleteChatRoomUnreadMessage`
These two functions are to set/delete the unreadMessages state inside the chatRoomLog.

##

So actually, I can say that we save two data in chatRoomLogs for each chatRoom. Those are `readUntil` and `unreadMessages`.
- The `readUntil` is used to set the `isNew` indicator position, and sometimes to set the initial scroll position. I think `isNew` indicator position should be same throughout all the devices for the same user, and it should be saved in KV store.
- The `unreadMessages` is used to save the new messages of different types; `MESSAGE_TYPES.TEXT`, `MESSAGE_TYPES.INTERACTIVE`, and `MESSAGE_TYPES.POLL`. It also used to calculate the number of unread messages which would be set the value of the badge for the chatroom.

You asked me if we should simplify the code by saving only `unreadMessages` which is a list of **only** message hashes.
- Even though, we would save only `unreadMessages`, I think the utility functions would have the same logic, and only two functions would be removed; `setChatRoomReadUntil` and `deleteChatRoomReadUntil`. But I think we should also save the `readUntil` in order to keep the same `isNew` indicator position throughout all the devices and `readUntil` is also used to delete `unreadMessages`.
- If we save only message hash in the `unreadMessages`, there would be problem to get to know their message type, and to delete some/all of them when user scrolls down and sees the messages.

So, in my honest opinion, I think all the functions are necessary, and the attributes of chatRoomLogs too.

##
Additionally, what I am going to say is that all the chatRoomLogs utility functions should be in a same invocation queue. That's because they could be run at the _same_(not exactly same, but with a few difference) time, and could set the wrong data.

Reply:
Thanks for that rundown @Silver-IT. I need to study your code more to come to a more informed conclusion.

However, I have some more questions:

> You asked me if we should simplify the code by saving only `unreadMessages` which is a list of only message hashes.

Sorry for not being clear, I actually meant what about saving only `readUntil`? I.e. only saving the hash of the message that we didn't read, and maybe (if necessary) saving a counter of the number of unread messages?

It seems like this code is saving unread messages in two different places:

1. In the KV store
2. In the contract Vuex state

But we should only be saving it in (2), the contract Vuex state.

Reply:
> Sorry for not being clear, I actually meant what about saving only `readUntil`? I.e. only saving the hash of the message that we didn't read, and maybe (if necessary) saving a counter of the number of unread messages?

I think there would be confusion from the names; `readUntil`, and `unreadMessages`.
Actually, `unreadMessages` doesn't store all the messages received after `readUntil` message. Just like in Slack, the red badge of the channel doesn't mean the number of all the unread messages. It is the number of mentions received after the last-seen message. `unreadMessages` stores only that kind of messages (The reason of that I can't call it `kind of mentions` is that we currently store three types of messages; TEXT, POLL, INTERACTIVE). And `readUntil` is the last-seen message and is used to display `isNew` indicator. So `readUntil` and `unreadMessages` are all needed, in my opinion.

> It would be awesome if we could just save it in one place (2), the contract Vuex state.

Yeah, we save chatRoomLogs in two places. Permanently in KV store, and temporarily in Vuex store. I guess you can understand why we should save them in KV store. The reasons why I think we need to save chatRoomLogs in Vuex store too are that we access the chatRoomLogs many times inside the Chat page and that I don't think we should access the KV store every time we need access. And also the another reason is that we have many vuex getters.
In the previous version, we did also save in two places; indexedStorage and Vuex store. The current version is similar but just exchanged indexedStorage with KV store.

Reply:
@Silver-IT I'm still not understanding. *Why* are you storing messages in the KV Store?

> unreadMessages stores only that kind of messages

Why?

> The current version is similar but just exchanged indexedStorage with KV store.

Why?

Can you not store messages in the KV Store?

The Vuex store should have only the ~20 or so most recent messages. The rest should be fetched using the `eventsAfter` API. And I see no reason to store messages in the KV Store... and so far no clear explanation for it

Reply:
> Why are you storing messages in the KV Store?

I think I made you confused again by using the word _messages_. Actually, we don't store messages in the KV store. Please check the following format of chatRoomLogs. `readUntil` has only three fields; `messageHash`, `createdHeight`, and possibly `deletedHeight`. And `unreadMessages` has three types too; `messageHash`, `type`, `createdHeight`. We don't save the whole message in the KV store.
```
chatRoomLogs: {
  [chatRoomID]: {
    readUntil: { messageHash, createdHeight, deletedHeight? },
    unreadMessages: [{ messageHash, type, createdHeight }]
  }
}
```

##
Also, it's true that we save at most 20 messages in each vuex store for each chatroom contract. We don't save `chatRoomLogs` in vuex store of the chatroom contract.

We save the `chatRoomLogs` in the KV store. Now whenever the user opens Chat page, they need to read the `readUntil` and `unreadMessages` of that chatroom. And also need to read many getters that uses `chatRoomLogs`. Here, if we don't save `chatRoomLogs` in vuex store, it's necessary to call `sbp('chelonia/kv/get', 
ourIdentityContractId, 'chatRoomLogs')` every time we access `chatRoomLogs`, and it's not good in terms of performance. So I think we should temporarily store `chatRoomLogs` in vuex store, and of course it's similar to how we save `loggedIn`/`lastLoggedIn`/`namespaceLookups`/`notifications`. Even though we have KV store for `lastLoggedIn`, we save `lastLoggedIn` in the vuex store for performance issue and convenience in using it.

Reply:
OK, thanks for helping clarify some things, but it's still not clear to me why you are storing the message hashes.

The only thing that's needed to display a red `------------------` unread messages line in the UI, plus the count of the unread messages, is the hash of the oldest unread message, plus a count of the number of unread messages.

That's it. So is it possible for you to store just that information, and nothing else?

Reply:
@taoeffect, I updated this PR and simplified the `chatRoomLogs` format by deleting `deletedHeight` of `readUntil`, `type` of `unreadMessages`. And also deleted the utility function `'gi.actions/identity/deleteChatRoomReadUntil'`.
```
// BEFORE
chatRoomLogs: {
  [chatRoomID]: {
    readUntil: { messageHash, createdHeight, deletedHeight? },
    unreadMessages: [{ messageHash, type, createdHeight }]
  }
}
// AFTER
chatRoomLogs: {
  [chatRoomID]: {
    readUntil: { messageHash, createdHeight },
    unreadMessages: [{ messageHash, createdHeight }]
  }
}
```

- **About `readUntil`**
  - **Why is `messageHash` needed?** In order to load events until that `readUntil.messageHash` when user opens the chat page.
  - **Why is `createdHeight` needed?** In order to set the `isNew` indicator position by comparing the heights of messsages to this `readUntil.createdHeight`.
- **About `unreadMessages`**
  - **Why is `unreadMessages` array? Why not save only the number of unread messages?** We can not save only the number of unread messages, because we need the following two fields.
  - **Why is `messageHash` needed?** Let's think about this case. user1 sends message `Hi @user2` in the chatroom and deletes that message. For user2, that message should be added to the `unreadMessages` array and also needs to be deleted again.Without the `messageHash` (and not saving that message hash in `unreadMessages` array or by only saving number of unread messages), we can not decide what to do with it. Should delete something from the list? Or should reduce the number of unread messages? Can not decide.
  - **Why is `createdHeight` needed?** Let's think about this case. user1 sends many messages with mentions to user2. Yeah, really many message of several pages. When user2 opens the chat page, what should he see? It's the last scroll position if it's the same device, or the `readUntil` message if it's the another device. At this moment, user2 haven't seen all the mentions because there are many messages of several pages. If the user2 scrolls down and checks messages one by one, the number of unread messages should be reduced one by one by comparing the heights of messages to the heights of unread messages.

##
If we really need to reduce the KV store, and the chatRoomLogs too, I have a solution. In the above `chatRoomLogs` format, we really need `createdHeight`. And we need `messageHash` either in the current version of project but it's not 100% necessary to save `messageHash` if we update the workflow. Now, we use `messageHash` in many places; to save chatRoomScrollPosition, to create message link, ... (Not sure for more)
If we replace them with the message height, we can also remove messageHash in the chatRoomLogs.
Here, what we need to consider is that based on the @corrideat idea, many messages could be created in one event. Many messages with different message hashes could have same height.

@taoeffect, if you agree on my suggestion to replace `messageHash` to `messageHeight` in all chatroom related data, I will create an issue and work on it soon. I think it could need some much changes so I would like to work it in another issue.

Reply:
Thank you for the explanation @Silver-IT.

> @taoeffect, if you agree on my suggestion to replace messageHash to messageHeight in all chatroom related data, I will create an issue and work on it soon.

That's OK, I think we can leave it as-is.

I will continue with the review now. One more note though (and I will add a comment on this elsewhere in the review), but I think `chatRoomLogs` needs to be renamed to something like `chatroomUnreadMessages` in order to be clearer and more accurate.

Reply:
Yeah, I will rename `chatRoomLogs` to `chatRoomUnreadMessages`.

However, it would have a field of `unreadMessages`, I think we should find a better name?
```
chatRoomUnreadMessages: {
  [chatRoomID]: {
    readUntil: { messageHash, createdHeight },
    unreadMessages: [{ messageHash, createdHeight }]
  }
}
```

Reply:
> However, it would have a field of `unreadMessages`, I think we should find a better name?

That seems fine to me, they are unread messages 🤷‍♂️